lasagne/layers/pool.py

```diff
@@ -690,18 +752,92 @@ def get_output_shape_for(self, input_shape):
-        a, b = self.scale_factor
+
+        # create weights for nearest neighbor interpolation
+        def nearest_neighbor_weight_matrix(input_dim, scale_factor):
+            indices = np.arange(0, input_dim * scale_factor)
+            source_indices = np.zeros((indices.shape[0]))
+
+            for x in indices:
+                source_indices[x] = np.min(
+                    np.array([np.floor(np.float(x) / scale_factor),
+                              input_dim - 1]))
+
+            weight_matrix = np.zeros((input_dim,
+                                      input_dim * scale_factor)
+                                     ).astype(theano.config.floatX)
+            source_indices = source_indices.astype(np.int)
+            for col in range(0, weight_matrix.shape[1]):
+                weight_matrix[source_indices[col], col] = 1
+
+            return weight_matrix
+
+        def _phase_shift(self, input, r):
+            def _ps(self, input, r):
+                bsize, _, a, b = self.input_shape
+                upscaled = T.reshape(input, (bsize, r, r, a, b))
+                upscaled = T.transpose(upscaled, (0, 3, 4, 2, 1))
+                upscaled = T.split(
+                    x=upscaled, splits_size=[1] * a, n_splits=a, axis=1)
+                upscaled = [T.reshape(x,
+                                      (bsize, b, r, r)) for x in upscaled]
+                upscaled = T.concatenate(upscaled, axis=2)
+                upscaled = T.split(
+                    x=upscaled, splits_size=[1] * b, n_splits=b, axis=1)
+                upscaled = [T.reshape(x, (bsize, a * r, r)) for x in upscaled]
+                upscaled = T.concatenate(upscaled, axis=2)
+                return upscaled.dimshuffle(0, 'x', 1, 2)
+
+            chan = int(self.input_shape[1] // (r ** 2))
+            splits = []
+            for q in range(chan):
+                splits.append(r ** 2)
+            if chan > 1:
+                Xc = T.split(
+                    x=input, splits_size=splits, n_splits=chan, axis=1)
+                upscaled = T.concatenate(
+                    [_ps(self, xc, r) for xc in Xc], axis=1)
+            else:
+                upscaled = _ps(self, input, r)
+            return upscaled
+
         upscaled = input
         if self.mode == 'repeat':
+            a, b = self.scale_factor
             if b > 1:
                 upscaled = T.extra_ops.repeat(upscaled, b, 3)
             if a > 1:
                 upscaled = T.extra_ops.repeat(upscaled, a, 2)
         elif self.mode == 'dilate':
+            a, b = self.scale_factor
             if b > 1 or a > 1:
                 output_shape = self.get_output_shape_for(input.shape)
                 upscaled = T.zeros(shape=output_shape, dtype=input.dtype)
                 upscaled = T.set_subtensor(upscaled[:, :, ::a, ::b], input)
+        elif self.mode == 'bilinear2D':
+            upscaled = T.nnet.abstract_conv.bilinear_upsampling(
+                input, self.scale_factor[0],
```

Comment:
I got notified by github that i didn't "submit" my comment. Usually, comment(shift + enter) alone is enough. The interface got complicated, I will check it out. Apologise if it has occurred as a repost.

Reply:
I'm not sure I understand your requested change- Theano handles creation of bilinear upsample filters within `bilinear_upsampling`, it's not an argument passed into `bilinear_upsampling` by the change in the PR.

The second input to `bilinear_upsampling` is the scalar scale factor to upsample the tensor by, not the filters used for convolution as it would be with something like `T.nnet.conv2d()`.

The output of `bilinear_upsampling` is a 4D tensor of dims `[batchsize, num_filters, row * scale_factor, col*scale_factor]` , where `num_filters` is the same as the number of filters of the previous layer- only the dims of the two trailing axes are affected by `bilinear_upsampling`.

Please do correct me if I'm wrong about any of the above, any feedback is welcome.

Reply:
> The second input to bilinear_upsampling is the scalar scale factor to upsample the tensor by, not the filters used for convolution as it would be with something like T.nnet.conv2d().

Yeah, my bad! Had it been the case, it should have been an issue in Theano primarily.